### DISCUSSION TITLE
websocket client basic authentication

### DISCUSSION THREAD

**Discussion by embetrix** [cesanta/mongoose#2227] [Ask questions, get answers] [answered]
I didn't find any information how the websocket client  basic authentication is working ?

for the http this works without any problem:

`curl -X POST http://192.168.178.101:8080  --digest -u "myuser:p!ass123" -w "%{http_code}"`

But for websocket for example using wscat :

`wscat -c ws://192.168.178.101:8080  --auth "myuser:p!ass123"`

doesn't work !

any hints are welcome

**Comment by scaprile** [comment, marked as answer]
There is no authentication in WS but in HTTP
It is _you_ who has to write the code to check for authentication and upgrade the connection to WS after authenticating the URL.
https://mongoose.ws/documentation/tutorials/device-dashboard/

https://github.com/cesanta/mongoose/blob/b379816178abdcd59135aa32f990a4b3bbbfb54b/examples/device-dashboard/net.c#L223-L225
https://github.com/cesanta/mongoose/blob/b379816178abdcd59135aa32f990a4b3bbbfb54b/examples/device-dashboard/net.c#L263-L265

https://mongoose.ws/documentation/tutorials/webui-login/

**Reply by Yhee2** [reply]
Hello, sorry to bother, but if I use Mongoose as a client to access an authenticated website via HTTP, how can I input the token? I would greatly appreciate it.

**Reply by cpq** [reply]
@Yhee2 depends on the auth mechanism that server uses.
Assuming it is Basic auth - then, a client must provide an `Authorization: Basic TOKEN`, where TOKEN is a base64-encoded `user:password` string. That header can be made like this:

```c
char buf[128];
size_t n = mg_snprintf(buf, sizeof(buf), "%s:%s", user, pass);  // Create user:password string
mg_ws_connect(mgr, url, fn, fn_data, "Authorization: Basic %M\r\n", mg_print_base64, n, buf);
```

**Reply by Yhee2** [reply]
Thank you very much for your answer.

**Comment by cpq** [comment]
@embetrix I've noticed you're using digest auth in curl argument. Mongoose can do basic auth, not digest.

Mongoose has function "creds" that fetches user and password from Authorization header and Cookie (access_token):
https://github.com/cesanta/mongoose/blob/7f3d450b3287297c61943f20f386c80877e244b8/examples/device-dashboard/net.c#L113-L114
